src/Snak/SnakObject.php

```diff
@@ -113,7 +113,34 @@ public function serialize() {
 	 * @param string $serialized
 	 */
 	public function unserialize( $serialized ) {
-		$this->propertyId = PropertyId::newFromNumber( unserialize( $serialized ) );
+		$this->propertyId = self::newPropertyId( $serialized );
+	}
+
+	/**
+	 * @param string $serialized
+	 *
+	 * @return PropertyId
+	 */
+	protected static function newPropertyId( $serialized ) {
+		if ( is_int( $serialized ) ) {
+			// numeric id, already unserialized
+			return PropertyId::newFromNumber( $serialized );
+		}
+
+		try {
+			// full ID, as a string, not serialized
+			return new PropertyId( $serialized );
+		} catch ( InvalidArgumentException $ex ) {
+			// noop
+		}
+
+		$unserialized = unserialize( $serialized );
+		if ( is_int( $unserialized ) ) {
```

Comment:
Sorry, but this code is weird. It does an `is_int` check twice? The reason is that this method is used for two very different use cases:
* In Value snaks this can be called with either `1` or `"P1"`.
* In NoValue and SomeValue snaks this can be called with `"i:42;"` or `"P1"`.

The problem is: there is no need to make these two use cases compatible with each other.